File: ruby_challenges/easy1/word_count.rb
class Phrase
  def initialize(phrase)
    @phrase = phrase
  end

  def word_count
    counts = {}
    words = @phrase.downcase.scan(/[\w']+/)

    words.each do |word|
      if word =~ /'.+'/
        word.delete_prefix!("'").delete_suffix!("'")
      end
      counts[word] = words.count(word)
    end

    counts
  end
end

# ALTERNATIVE SOLUTION:

# (using #split instead of #scan)
# class Phrase
#   def initialize(phrase)
#     @phrase = phrase
#   end

#   def word_count
#     counts = {}
#     words = @phrase.downcase.split(/[^\w']+/)

#     words.each do |word|
#       if word =~ /'.+'/
#         word.delete_prefix!("'").delete_suffix!("'")
#       end
#       counts[word] = words.count(word)
#     end
    
#     counts
#   end
# end

File: js210/lesson_5/wordCount.js
function wordCount(string) {
  const words = string.split(' ');
  const result = {};

  for (let index = 0; index < words.length; index++) {
    let word = words[index];

    if (!result[word]) {
      let regexp = RegExp(`${word}`, 'g');
      let count = string.match(regexp).length;
      result[word] = count;
    }
  }

  return result;
}

console.log(wordCount('box car cat bag box'));  // { box: 2, car: 1, cat: 1, bag: 1 }


// ALTERNATIVE SOLUTION:

// function wordCount(text) {
//   let count = {};
//   let words = text.split(' ');

//   for (let index = 0; index < words.length; index += 1) {
//     let word = words[index];
//     if (!count[word]) {
//       count[word] = 0;
//     }

//     count[word] += 1;
//   }

//   return count;
// }
